File: ios/MyApp/Models/Task.swift
import Foundation

enum TaskPriority: String, Codable, CaseIterable, Identifiable, Sendable {
    case low
    case medium
    case high

    var id: String { rawValue }

    var label: String {
        switch self {
        case .low: "Low"
        case .medium: "Medium"
        case .high: "High"
        }
    }
}

enum TaskStatus: String, Codable, CaseIterable, Identifiable, Sendable {
    case pending
    case inProgress = "in_progress"
    case completed

    var id: String { rawValue }

    var label: String {
        switch self {
        case .pending: "Pending"
        case .inProgress: "In Progress"
        case .completed: "Completed"
        }
    }
}

enum TaskProductivity: String, Codable, CaseIterable, Identifiable, Sendable {
    case fast
    case moderate
    case slow

    var id: String { rawValue }

    var label: String {
        switch self {
        case .fast: "Fast"
        case .moderate: "Moderate"
        case .slow: "Slow"
        }
    }
}

struct ChecklistItem: Codable, Hashable, Sendable {
    var text: String
    var done: Bool

    init(text: String, done: Bool = false) {
        self.text = text
        self.done = done
    }
}

/// A time override and/or completion marker for a single occurrence of a
/// repeating task. An occurrence may carry a time override, a per-date
/// completion marker, or both.
struct RepeatOverride: Codable, Hashable, Sendable {
    var startAt: Date?
    var endAt: Date?
    var completed: Bool? = nil

    private enum CodingKeys: String, CodingKey {
        case startAt = "start_at"
        case endAt = "end_at"
        case completed
    }
}

enum OccurrenceScope: String, Sendable {
    case thisEventOnly = "this_event_only"
    case fromNowOnwards = "from_now_onwards"
}

/// Formats dates into the `yyyy-MM-dd` keys used by the backend's
/// `repeat_overrides` map. Uses the device's calendar day so an override lines
/// up with the occurrence the user tapped, regardless of UTC offset.
enum OccurrenceDateKey {
    private static let formatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = .current
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter
    }()

    static func key(for date: Date) -> String {
        formatter.string(from: date)
    }
}

struct TaskItem: Codable, Identifiable, Hashable, Sendable {
    let id: UUID
    let userId: UUID
    var title: String
    var description: String?
    var deadline: Date?
    var startAt: Date?
    var endAt: Date?
    var priority: TaskPriority
    var status: TaskStatus
    var estimatedDuration: Int?
    var actualDuration: Int?
    var productivity: TaskProductivity?
    var startedAt: Date?
    var completedAt: Date?
    var category: String?
    var notes: String?
    var checklist: [ChecklistItem]?
    var repeatWeekdays: [Int]?
    var repeatEndsOn: Date?
    var repeatOverrides: [String: RepeatOverride]? = nil
    var beforeTaskIds: [UUID]?
    var afterTaskIds: [UUID]?
    var isArchived: Bool
    var progressPercent: Int
    var createdAt: Date
    var updatedAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case userId = "user_id"
        case title
        case description
        case deadline
        case startAt = "start_at"
        case endAt = "end_at"
        case priority
        case status
        case estimatedDuration = "estimated_duration"
        case actualDuration = "actual_duration"
        case productivity
        case startedAt = "started_at"
        case completedAt = "completed_at"
        case category
        case notes
        case checklist
        case repeatWeekdays = "repeat_weekdays"
        case repeatEndsOn = "repeat_ends_on"
        case repeatOverrides = "repeat_overrides"
        case beforeTaskIds = "before_task_ids"
        case afterTaskIds = "after_task_ids"
        case isArchived = "is_archived"
        case progressPercent = "progress_percent"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }
}

struct TaskListResponse: Codable, Sendable {
    let items: [TaskItem]
    let total: Int
}

struct TaskParseRequest: Encodable, Sendable {
    let text: String
    let timezone: String
}

struct TaskCreateRequest: Codable, Sendable {
    let title: String
    let description: String?
    let deadline: Date?
    let startAt: Date?
    let endAt: Date?
    let priority: TaskPriority
    let status: TaskStatus
    let estimatedDuration: Int?
    let category: String?
    let notes: String?
    let checklist: [ChecklistItem]?
    let repeatWeekdays: [Int]?
    let repeatEndsOn: Date?
    let beforeTaskIds: [UUID]?
    let afterTaskIds: [UUID]?

    init(
        title: String,
        description: String?,
        deadline: Date?,
        startAt: Date?,
        endAt: Date?,
        priority: TaskPriority,
        status: TaskStatus,
        estimatedDuration: Int?,
        category: String?,
        notes: String?,
        repeatWeekdays: [Int]?,
        beforeTaskIds: [UUID]? = nil,
        afterTaskIds: [UUID]? = nil,
        repeatEndsOn: Date? = nil,
        checklist: [ChecklistItem]? = nil
    ) {
        self.title = title
        self.description = description
        self.deadline = deadline
        self.startAt = startAt
        self.endAt = endAt
        self.priority = priority
        self.status = status
        self.estimatedDuration = estimatedDuration
        self.category = category
        self.notes = notes
        self.checklist = checklist
        self.repeatWeekdays = repeatWeekdays
        self.beforeTaskIds = beforeTaskIds
        self.afterTaskIds = afterTaskIds
        self.repeatEndsOn = repeatEndsOn
    }

    init(from local: LocalTask) {
        title = local.title
        description = local.taskDescription
        deadline = local.deadline
        startAt = local.startAt
        endAt = local.endAt
        priority = TaskPriority(rawValue: local.priorityRaw) ?? .medium
        status = TaskStatus(rawValue: local.statusRaw) ?? .pending
        estimatedDuration = local.estimatedDuration
        category = local.category
        notes = local.notes
        checklist = local.checklist
        repeatWeekdays = local.repeatWeekdays
        beforeTaskIds = local.beforeTaskIds
        afterTaskIds = local.afterTaskIds
        repeatEndsOn = local.repeatEndsOn
    }

    enum CodingKeys: String, CodingKey {
        case title
        case description
        case deadline
        case startAt = "start_at"
        case endAt = "end_at"
        case priority
        case status
        case estimatedDuration = "estimated_duration"
        case category
        case notes
        case checklist
        case repeatWeekdays = "repeat_weekdays"
        case repeatEndsOn = "repeat_ends_on"
        case beforeTaskIds = "before_task_ids"
        case afterTaskIds = "after_task_ids"
    }
}

struct TaskUpdateRequest: Encodable, Sendable {
    let title: String
    let description: String?
    let deadline: Date?
    let startAt: Date?
    let endAt: Date?
    let priority: TaskPriority
    let status: TaskStatus
    let estimatedDuration: Int?
    let actualDuration: Int?
    let category: String?
    let notes: String?
    let checklist: [ChecklistItem]?
    let repeatWeekdays: [Int]?
    let repeatEndsOn: Date?
    let repeatOverrides: [String: RepeatOverride]?
    let beforeTaskIds: [UUID]?
    let afterTaskIds: [UUID]?

    init(task: TaskItem) {
        title = task.title
        description = task.description
        deadline = task.deadline
        startAt = task.startAt
        endAt = task.endAt
        priority = task.priority
        status = task.status
        estimatedDuration = task.estimatedDuration
        actualDuration = task.actualDuration
        category = task.category
        notes = task.notes
        checklist = task.checklist
        repeatWeekdays = task.repeatWeekdays
        repeatEndsOn = task.repeatEndsOn
        repeatOverrides = task.repeatOverrides
        beforeTaskIds = task.beforeTaskIds
        afterTaskIds = task.afterTaskIds
    }

    init(from local: LocalTask) {
        title = local.title
        description = local.taskDescription
        deadline = local.deadline
        startAt = local.startAt
        endAt = local.endAt
        priority = TaskPriority(rawValue: local.priorityRaw) ?? .medium
        status = TaskStatus(rawValue: local.statusRaw) ?? .pending
        estimatedDuration = local.estimatedDuration
        actualDuration = local.actualDuration
        category = local.category
        notes = local.notes
        checklist = local.checklist
        repeatWeekdays = local.repeatWeekdays
        repeatEndsOn = local.repeatEndsOn
        repeatOverrides = local.repeatOverridesData.flatMap {
            try? JSONCoding.decoder.decode(
                [String: RepeatOverride].self,
                from: $0
            )
        }
        beforeTaskIds = local.beforeTaskIds
        afterTaskIds = local.afterTaskIds
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(title, forKey: .title)
        try container.encode(description, forKey: .description)
        try container.encode(deadline, forKey: .deadline)
        try container.encode(startAt, forKey: .startAt)
        try container.encode(endAt, forKey: .endAt)
        try container.encode(priority, forKey: .priority)
        try container.encode(status, forKey: .status)
        try container.encode(estimatedDuration, forKey: .estimatedDuration)
        try container.encode(actualDuration, forKey: .actualDuration)
        try container.encode(category, forKey: .category)
        try container.encode(notes, forKey: .notes)
        try container.encode(checklist, forKey: .checklist)
        try container.encode(repeatWeekdays, forKey: .repeatWeekdays)
        try container.encode(repeatEndsOn, forKey: .repeatEndsOn)
        try container.encodeIfPresent(repeatOverrides, forKey: .repeatOverrides)
        try container.encodeIfPresent(beforeTaskIds, forKey: .beforeTaskIds)
        try container.encodeIfPresent(afterTaskIds, forKey: .afterTaskIds)
    }

    private enum CodingKeys: String, CodingKey {
        case title
        case description
        case deadline
        case startAt = "start_at"
        case endAt = "end_at"
        case priority
        case status
        case estimatedDuration = "estimated_duration"
        case actualDuration = "actual_duration"
        case category
        case notes
        case checklist
        case repeatWeekdays = "repeat_weekdays"
        case repeatEndsOn = "repeat_ends_on"
        case repeatOverrides = "repeat_overrides"
        case beforeTaskIds = "before_task_ids"
        case afterTaskIds = "after_task_ids"
    }
}

struct CompleteTaskRequest: Encodable, Sendable {
    let actualMinutes: Int?
    let productivity: TaskProductivity?
    let occurrenceDate: String?
    let timezone: String?

    private enum CodingKeys: String, CodingKey {
        case actualMinutes = "actual_minutes"
        case productivity
        case occurrenceDate = "occurrence_date"
        case timezone
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encodeIfPresent(actualMinutes, forKey: .actualMinutes)
        try container.encodeIfPresent(productivity, forKey: .productivity)
        try container.encodeIfPresent(occurrenceDate, forKey: .occurrenceDate)
        try container.encodeIfPresent(timezone, forKey: .timezone)
    }
}

struct RecordTimeRequest: Encodable, Sendable {
    let minutes: Int

    private enum CodingKeys: String, CodingKey {
        case minutes = "actual_duration"
    }
}

struct RescheduleRequest: Encodable, Sendable {
    let minutesRemaining: Int
    let reason: String?
    let timezone: String
    let deadline: Date?

    private enum CodingKeys: String, CodingKey {
        case minutesRemaining = "minutes_remaining"
        case reason
        case timezone
        case deadline
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(minutesRemaining, forKey: .minutesRemaining)
        try container.encodeIfPresent(reason, forKey: .reason)
        try container.encode(timezone, forKey: .timezone)
        try container.encodeIfPresent(deadline, forKey: .deadline)
    }
}

struct RescheduleResponse: Codable, Sendable {
    let task: TaskItem
    let blocks: [CalendarBlock]
}

struct OccurrenceUpdateRequest: Encodable, Sendable {
    let date: String
    let scope: String
    let startAt: Date
    let endAt: Date
    let timezone: String

    private enum CodingKeys: String, CodingKey {
        case date
        case scope
        case startAt = "start_at"
        case endAt = "end_at"
        case timezone
    }
}

struct OccurrenceUpdateResponse: Codable, Sendable {
    let task: TaskItem
    let newTask: TaskItem?

    private enum CodingKeys: String, CodingKey {
        case task
        case newTask = "new_task"
    }
}

File: backend/app/schemas/task.py
import uuid
from datetime import date
from datetime import datetime
from typing import Literal

from pydantic import BaseModel
from pydantic import ConfigDict
from pydantic import Field
from pydantic import field_validator
from pydantic import model_validator

from app.models.task import TaskPriority
from app.models.task import TaskProductivity
from app.models.task import TaskStatus
from app.schemas.calendar import CalendarBlockResponse


def _normalize_weekdays(value: list[int] | None) -> list[int] | None:
    """Validate weekday numbers (0=Sunday..6=Saturday) and dedupe them."""
    if not value:
        return None
    if any(day < 0 or day > 6 for day in value):
        raise ValueError(
            "weekdays must be integers between 0 (Sunday) and 6 (Saturday)"
        )
    return sorted(set(value))


class ChecklistItem(BaseModel):
    text: str = Field(min_length=0, max_length=500)
    done: bool = False


class TaskCreate(BaseModel):
    title: str = Field(min_length=1, max_length=255)
    description: str | None = None
    deadline: datetime | None = None
    start_at: datetime | None = None
    end_at: datetime | None = None
    priority: TaskPriority = TaskPriority.medium
    status: TaskStatus = TaskStatus.pending
    estimated_duration: int | None = Field(default=None, ge=1, le=525600)
    actual_duration: int | None = Field(default=None, ge=0, le=525600)
    productivity: TaskProductivity | None = None
    category: str | None = Field(default=None, max_length=100)
    notes: str | None = None
    checklist: list[ChecklistItem] | None = None
    repeat_weekdays: list[int] | None = None
    repeat_ends_on: datetime | None = None
    before_task_ids: list[uuid.UUID] | None = None
    after_task_ids: list[uuid.UUID] | None = None

    @field_validator("title")
    @classmethod
    def title_not_blank(cls, value: str) -> str:
        if not value.strip():
            raise ValueError("title must not be blank")
        return value.strip()

    @field_validator("repeat_weekdays")
    @classmethod
    def repeat_weekdays_valid(cls, value: list[int] | None) -> list[int] | None:
        return _normalize_weekdays(value)

    @model_validator(mode="after")
    def fixed_event_times_valid(self):
        if self.start_at is None and self.end_at is None:
            return self
        if self.start_at is None or self.end_at is None:
            raise ValueError("start_at and end_at must be set together")
        if self.end_at <= self.start_at:
            raise ValueError("end_at must be after start_at")
        return self


class RepeatOverride(BaseModel):
    # A repeating occurrence can hold a time override (start_at/end_at) and/or
    # a per-date completion marker. Either may be present on its own.
    start_at: datetime | None = None
    end_at: datetime | None = None
    completed: bool = False
    actual_duration: int | None = Field(default=None, ge=0, le=525600)

    @model_validator(mode="after")
    def times_valid(self):
        if self.start_at is not None and self.end_at is not None:
            if self.end_at <= self.start_at:
                raise ValueError("end_at must be after start_at")
        return self


class TaskUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = None
    deadline: datetime | None = None
    start_at: datetime | None = None
    end_at: datetime | None = None
    priority: TaskPriority | None = None
    status: TaskStatus | None = None
    estimated_duration: int | None = Field(default=None, ge=1, le=525600)
    actual_duration: int | None = Field(default=None, ge=0, le=525600)
    productivity: TaskProductivity | None = None
    category: str | None = Field(default=None, max_length=100)
    notes: str | None = None
    checklist: list[ChecklistItem] | None = None
    repeat_weekdays: list[int] | None = None
    repeat_ends_on: datetime | None = None
    repeat_overrides: dict[str, RepeatOverride] | None = None
    before_task_ids: list[uuid.UUID] | None = None
    after_task_ids: list[uuid.UUID] | None = None

    @field_validator("title")
    @classmethod
    def title_not_blank(cls, value: str | None) -> str | None:
        if value is not None and not value.strip():
            raise ValueError("title must not be blank")
        return value.strip() if value is not None else None

    @field_validator("repeat_weekdays")
    @classmethod
    def repeat_weekdays_valid(cls, value: list[int] | None) -> list[int] | None:
        return _normalize_weekdays(value)

    @field_validator("before_task_ids")
    @classmethod
    def before_ids_valid(cls, value: list[uuid.UUID] | None) -> list[uuid.UUID] | None:
        return list(dict.fromkeys(value)) if value else None

    @field_validator("after_task_ids")
    @classmethod
    def after_ids_valid(cls, value: list[uuid.UUID] | None) -> list[uuid.UUID] | None:
        return list(dict.fromkeys(value)) if value else None

    @model_validator(mode="after")
    def fixed_event_times_valid(self):
        if self.start_at is not None and self.end_at is not None:
            if self.end_at <= self.start_at:
                raise ValueError("end_at must be after start_at")
        return self


class TaskResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    user_id: uuid.UUID
    title: str
    description: str | None
    deadline: datetime | None
    start_at: datetime | None
    end_at: datetime | None
    priority: TaskPriority
    status: TaskStatus
    estimated_duration: int | None
    actual_duration: int | None
    productivity: TaskProductivity | None
    started_at: datetime | None
    completed_at: datetime | None
    category: str | None
    notes: str | None
    checklist: list[ChecklistItem] | None
    progress_percent: int
    repeat_weekdays: list[int] | None
    repeat_ends_on: datetime | None
    repeat_overrides: dict[str, RepeatOverride] | None
    before_task_ids: list[uuid.UUID] | None
    after_task_ids: list[uuid.UUID] | None
    is_archived: bool
    created_at: datetime
    updated_at: datetime


class TaskListResponse(BaseModel):
    items: list[TaskResponse]
    total: int


class CompleteTaskRequest(BaseModel):
    actual_minutes: int | None = Field(default=None, ge=1, le=525600)
    productivity: TaskProductivity | None = None
    # When supplied for a repeating task, completion applies only to the
    # occurrence on that local date (the series keeps going). Without it,
    # repeating tasks complete today's occurrence; non-repeating tasks are
    # completed as a whole.
    occurrence_date: date | None = None
    timezone: str = "UTC"


class SnoozeRequest(BaseModel):
    minutes: int = Field(ge=1, le=1440)
    timezone: str = "UTC"


class SnoozeResponse(BaseModel):
    task: TaskResponse
    blocks: list[CalendarBlockResponse]


class RescheduleRequest(BaseModel):
    minutes_remaining: int = Field(ge=1, le=525600)
    reason: str | None = Field(default=None, max_length=2000)
    timezone: str = "UTC"
    deadline: datetime | None = None


class RescheduleResponse(BaseModel):
    task: TaskResponse
    blocks: list[CalendarBlockResponse]


class OccurrenceUpdateRequest(BaseModel):
    date: date
    scope: Literal["this_event_only", "from_now_onwards"]
    start_at: datetime
    end_at: datetime
    timezone: str = "UTC"

    @model_validator(mode="after")
    def times_valid(self):
        if self.end_at <= self.start_at:
            raise ValueError("end_at must be after start_at")
        return self


class OccurrenceUpdateResponse(BaseModel):
    task: TaskResponse
    new_task: TaskResponse | None = None


class TaskParseRequest(BaseModel):
    text: str = Field(min_length=1, max_length=2000)
    timezone: str = "UTC"


class TaskParseResponse(BaseModel):
    title: str | None = None
    description: str | None = None
    deadline: datetime | None = None
    start_at: datetime | None = None
    end_at: datetime | None = None
    priority: TaskPriority | None = None
    estimated_duration: int | None = Field(default=None, ge=1, le=525600)
    category: str | None = None
    notes: str | None = None
